### PR TITLE
[TO_REVIEW] Upgrade DANN

### PR DESCRIPTION
**Pull request by tgnassou**
- Add a new domain classifier for DANN
- Put different lr for domain classifier en base_module

### CODE REVIEW

**Comment by antoinedemathelin** [comment]
Ca m'a l'air Ok.
Mais du coup le learning rate du end-classifier sera le même que l'encodeur.
En général, on en prend un plus grand. Après, si on préentraine le end-classifieur sur le source only, ca peut le faire

File: solvers/dann.py [hunk 1]
```diff
@@ -51,8 +53,12 @@ def get_estimator(self, n_classes, device, dataset_name, **kwargs):
             device=device,
             callbacks=[params['lr_scheduler']],
             max_epochs=params['max_epochs'],
-            lr=params['lr'],
             num_features=params['num_features'],
+            domain_classifier=DomainClassifier(num_features=params['num_features']),
```
Comment (antoinedemathelin) [inline solvers/dann.py:57]: Cf. commentaire

**Comment by YanisLalou** [comment]
> Ca m'a l'air Ok.
> Mais du coup le learning rate du end-classifier sera le même que l'encodeur.
> En général, on en prend un plus grand. Après, si on préentraine le end-classifieur sur le source only, ca peut le faire

That's the thing, we don't pretrain the model rn. I know there's a pending PR on skada but it hasn't been tested yet

**Comment by antoinedemathelin** [comment]
Maybe we can define another learning rate for the end classifier in the dictionary ?
It should be possible by giving a name to the end-classifier module?
The end-classifier will be randomly initialized, so it can takes many steps before it reach a good solution if we train it at the rate of the encoder.
The risk is to change all the encoder instead of fitting the classifier

**Review by YanisLalou — APPROVED** [review]
lgtm

File: benchmark_utils/backbones_architecture.py [hunk 2]
```diff
@@ -98,5 +96,72 @@ def __init__(self, n_chans, n_classes, input_window_samples):
...
+        super().__init__()
+        self.classifier = nn.Sequential(
+            nn.Linear(num_features, hidden_size),
+            nn.ReLU(),
+            nn.Dropout(0.5),
+            nn.Linear(hidden_size, hidden_size),
+            nn.ReLU(),
+            nn.Dropout(0.5),
+            nn.Linear(hidden_size, n_classes),
+            nn.Sigmoid(),
```
Comment (antoinecollas) [inline benchmark_utils/backbones_architecture.py:151]: Why a sigmoid? e.g. `ShallowConvNet` returns logits.

Comment (YanisLalou) [reply benchmark_utils/backbones_architecture.py:151]: Its the domain classifier. It does Binary classification between the Target and Source domain. That's why using a sigmoid makes sense

Comment (antoinecollas) [reply benchmark_utils/backbones_architecture.py:151]: Ok, I just saw we use `torch.nn.BCELoss()` in skada so the pr looks good to me!

**Review by antoinedemathelin — APPROVED** [review]
